### PR TITLE
Hero: Improve pattern quality

### LINKED ISSUE #321: Hero: Improve pattern quality
Hero background pattern images are shown in lower quality than expected

Ref: https://adobe-dx-support.slack.com/archives/C06H8UNQ3RS/p1710760013243169

Ideas:
- For pattern background images, use format webpll (loss-less) instead of webply (lossy). But don't use this format for images (see comments by @rootpea below)

### PR DESCRIPTION
**Pull request by pahupe**
By specifying 'full background image' for a hero, one can request a loss-less rendition.

fix: 'full background image' will force the image to be rendered in loss-less quality (webpll)

Fix #321

Test URLs (hero will change its quality):
- Before: https://main--hlx-test--urfuwo.hlx.live/draft/hupe/321-pattern-quality
- After: https://321-hero-improve-pattern-quality--hlx-test--urfuwo.hlx.live/draft/hupe/321-pattern-quality

### CODE REVIEW

File: aemedge/blocks/hero/hero.js [hunk 1]
```diff
@@ -78,6 +78,9 @@ export default async function decorate(block) {
   // get images for background
   let picture = block.querySelector(':scope div > div > picture');
   if (picture) {
+    if (block.classList.contains('pattern')) {
+      picture.innerHTML = picture.innerHTML.replaceAll('webply', 'webpll');
```
Comment (mhaack) [inline aemedge/blocks/hero/hero.js:82]: Should we only replace it for the right src attributes or the picture & img tags?

File: aemedge/blocks/hero/hero.js [hunk 2]
```diff
@@ -78,6 +78,9 @@ export default async function decorate(block) {
   // get images for background
   let picture = block.querySelector(':scope div > div > picture');
   if (picture) {
+    if (block.classList.contains('pattern')) {
```
Comment (mhaack) [inline aemedge/blocks/hero/hero.js:81]: No, this should be either for all background image or nothing, the author should not care about this.

**Review by mhaack — CHANGES_REQUESTED** [review]

File: aemedge/blocks/hero/hero.js (hunk 2 above)
Comment (pahupe) [reply aemedge/blocks/hero/hero.js]: Fixed.

File: aemedge/blocks/hero/hero.js (hunk 1 above)
Comment (karlpauls) [reply aemedge/blocks/hero/hero.js:82]: I think we should either enhance createOptimizedPicture to accept a format hint between the two and then replace the picture with the result of createOptimizedPicture call or at a minimum, we should iterate over the source elements and only replace the scrset if the type is image/webpl

**Review by karlpauls — CHANGES_REQUESTED** [review]

File: aemedge/blocks/hero/hero.js (hunk 1 above)
Comment (pahupe) [reply aemedge/blocks/hero/hero.js:91]: > Should we only replace it for the right src attributes or the picture & img tags?

I had a look at the server-side generated block, and this is what I got:
image

I believe the relevant part is not the source type="...", but rather the "format=webply" parameter in the URL. As you can see, there's also other entries with "format=png" that I don't (want to) touch.

I therefore would suggest to replace "format=webply" with "format=webpll", as the decision to generate the target format "webply" (= lossy webp) is made by the EDS server-side code, and we only want to correct that

Comment (pahupe) [reply aemedge/blocks/hero/hero.js:91]: > I think we should either enhance createOptimizedPicture to accept a format hint between the two and then replace the picture with the result of createOptimizedPicture call or at a minimum, we should iterate over the source elements and only replace the scrset if the type is image/webpl

I don't fully understand the approach. The <picture> element is generated server-side, so our JS code (createOptimizedPicture) does not generate it.

Can you please detail your approach, @karlpauls ?

**Review by mhaack — APPROVED** [review]

**Comment by pahupe** [comment]
@karlpauls : Integrated your requested change

**Review by karlpauls — APPROVED** [review]

**Pull request merged**